File: spark-scala/src/main/scala/rustfs/Rustfs.scala
package rustfs

import software.amazon.awssdk.auth.credentials._
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.s3._
import software.amazon.awssdk.services.s3.model._
import io.github.cdimascio.dotenv.Dotenv

import java.net.URI
import scala.jdk.CollectionConverters._

final case class S3Object(key: String, size: Long)

object Env {
  private val dotenv = Dotenv.configure().ignoreIfMissing().load()

  def get(name: String, default: => String = null): String = {
    val v = Option(System.getenv(name))
      .orElse(Option(dotenv.get(name, null)))
      .map(_.trim)
      .filter(_.nonEmpty)
      .orNull
    if (v != null) v
    else if (default != null) default
    else throw new IllegalArgumentException(s"Missing env var: $name")
  }
}

final class RustFS(
                    bucket: String = Env.get("AWS_S3_BUCKET"),
                    endpoint: String = Env.get("AWS_S3_ENDPOINT", "http://localhost:9000"),
                    region: String = Env.get("AWS_REGION", "us-east-1"),
                    accessKeyId: String = Env.get("AWS_ACCESS_KEY_ID"),
                    secretAccessKey: String = Env.get("AWS_SECRET_ACCESS_KEY")
                  ) {
  private val client: S3Client =
    S3Client.builder()
      .endpointOverride(URI.create(endpoint))
      .region(Region.of(region))
      .credentialsProvider(
        StaticCredentialsProvider.create(
          AwsBasicCredentials.create(accessKeyId, secretAccessKey)
        )
      )
      .forcePathStyle(true)
      .build()

  def list(prefix: String = ""): Seq[S3Object] = {
    val req = ListObjectsV2Request.builder().bucket(bucket).prefix(prefix).build()
    client.listObjectsV2Paginator(req).contents().asScala.map(o => S3Object(o.key(), o.size())).toSeq
  }

  def read(key: String): Array[Byte] = {
    val req = GetObjectRequest.builder().bucket(bucket).key(key).build()
    client.getObjectAsBytes(req).asByteArray()
  }

  def put(key: String, bytes: Array[Byte], contentType: String = "application/octet-stream"): Unit = {
    val req = PutObjectRequest.builder().bucket(bucket).key(key).contentType(contentType).build()
    client.putObject(req, software.amazon.awssdk.core.sync.RequestBody.fromBytes(bytes))
    ()
  }
}

File: python/server/rustfs.py
import os
from dataclasses import dataclass
from typing import Optional

import boto3
from botocore.client import Config


@dataclass(frozen=True)
class RustFsConfig:
    bucket: str
    endpoint_url: Optional[str]
    region: str
    access_key_id: str
    secret_access_key: str

# Rustfs has the endpoint public that will be in front of nginx

class RustFs:

    def __init__(self, cfg: RustFsConfig) -> None:
        self.cfg = cfg
        print(cfg)
        self.s3 = boto3.client(
            "s3",
            region_name=cfg.region,
            aws_access_key_id=cfg.access_key_id,
            aws_secret_access_key=cfg.secret_access_key,
            endpoint_url=cfg.endpoint_url,
            config=Config(signature_version="s3v4", s3={"addressing_style": "path"}),
        )

    def presigned_put_url(
        self,
        *,
        key: str,
        expires_in_seconds: int = 900,
        content_type: str = "application/octet-stream",
    ) -> str:
        return self.s3.generate_presigned_url(
            ClientMethod="put_object",
            Params={
                "Bucket": self.cfg.bucket,
                "Key": key,
                "ContentType": content_type,
            },
            ExpiresIn=expires_in_seconds,
        )


def load_rustfs_from_env() -> RustFs:
    def req(name: str) -> str:
        v = os.getenv(name)
        if not v:
            raise RuntimeError(f"Missing env var: {name}")
        return v

    cfg = RustFsConfig(
        bucket=req("AWS_S3_BUCKET"),
        endpoint_url=os.getenv("AWS_S3_PUBLIC_ENDPOINT"),
        region=os.getenv("AWS_REGION", "us-east-1"),
        access_key_id=req("AWS_ACCESS_KEY_ID"),
        secret_access_key=req("AWS_SECRET_ACCESS_KEY"),
    )
    return RustFs(cfg)